db/compaction/compaction_outputs.cc

```diff
@@ -190,13 +190,60 @@ bool CompactionOutputs::ShouldStopBefore(const CompactionIterator& c_iter) {
   // etc.
   const Slice& internal_key = c_iter.key();
   const uint64_t previous_overlapped_bytes = grandparent_overlapped_bytes_;
-  size_t num_grandparent_boundaries_crossed =
-      UpdateGrandparentBoundaryInfo(internal_key);
+  const InternalKeyComparator* icmp =
+      &compaction_->column_family_data()->internal_comparator();
+  // if compaction_->output_level() == 0, there is no need to update grandparent
+  // info, and that `grandparent` should be empty.
+  size_t num_grandparent_boundaries_crossed = 0;
+  bool should_stop_for_ttl = false;
+  if (compaction_->output_level() > 0) {
+    num_grandparent_boundaries_crossed =
+        UpdateGrandparentBoundaryInfo(internal_key);
+    // TTL state should be updated for each key
+    if (!files_to_cut_for_ttl_.empty()) {
+      if (cur_files_to_cut_for_ttl_ != -1) {
+        // Previous key is inside the range of a file
+        if (icmp->Compare(internal_key,
+                          files_to_cut_for_ttl_[cur_files_to_cut_for_ttl_]
+                              ->largest.Encode()) > 0) {
+          next_files_to_cut_for_ttl_ = cur_files_to_cut_for_ttl_ + 1;
+          cur_files_to_cut_for_ttl_ = -1;
+          should_stop_for_ttl = true;
+        }
+      } else {
+        // Look for the key position
+        while (next_files_to_cut_for_ttl_ <
+               static_cast<int>(files_to_cut_for_ttl_.size())) {
+          if (icmp->Compare(internal_key,
+                            files_to_cut_for_ttl_[next_files_to_cut_for_ttl_]
+                                ->smallest.Encode()) >= 0) {
+            if (icmp->Compare(internal_key,
+                              files_to_cut_for_ttl_[next_files_to_cut_for_ttl_]
+                                  ->largest.Encode()) <= 0) {
+              // With in the current file
+              cur_files_to_cut_for_ttl_ = next_files_to_cut_for_ttl_;
+              should_stop_for_ttl = true;
+              break;
+            }
+            // Beyond the current file
+            next_files_to_cut_for_ttl_++;
+          } else {
+            // Still fall into the gap
+            break;
+          }
+        }
+      }
+    }
```

Comment:
Since you are already on this, could you refactor this into a function Update... like UpdateGrandparentBoundaryInfo to prevent future developers adding return statement before that, though I could see your "// TTL state should be updated for each key" is trying to do so.  I could see previous efforts in trying to ensure object states are correctly updated in face of early return https://github.com/facebook/rocksdb/commit/f3cc66632b9e9fda4822e7512beb523d78179205#diff-82239556eaea03c0bf21ed100ccc839551faec954c5f05843850b1b05e835802R145-R146.

Reply:
Updated with a new function to update ttl states: UpdateFilesToCutForTTLStates().